docs/userguide/design.md

```diff
@@ -206,35 +206,84 @@ service objects, organized according to locale. Then, if a particular locale's
+its data. It can be used to provide an early indication of configuration or
+installation problems with ICU data files. Its use is optional; it is not
+required for the correct functioning of ICU.
 
 When an application is terminating, it may optionally call the function
-`u_cleanup(void)`, which will free any heap storage that has been allocated and
-held by the ICU library. The main benefit of `u_cleanup()` occurs when using
-memory leak checking tools while debugging or testing an application. Without
-`u_cleanup()`, memory being held by the ICU library will be reported as leaks.
+`u_cleanup()`, which frees all heap storage and other system resources that
+are held internally by the ICU library. The main benefit of `u_cleanup()` occurs
+when using memory leak checking tools while testing an application. Without
+`u_cleanup()`, resources being held by ICU library will be reported as leaking.
+
+Before calling `u_cleanup()`, all ICU objects that were created by the
+application must be deleted, and all ICU services (plain C APIs) must be closed.
+
+For some platforms the configure option `--enable-auto-cleanup`, or defining
+the option `UCLN_NO_AUTO_CLEANUP` to 0, will add code which automatically cleans
+up ICU when its shared library is unloaded. See comments in `ucln_imp.h`
+
+#### C++ Static Initialization and Destruction
+
+The ICU library itself does not rely on C++ static initializers, meaning that
+applications will not encounter order-of-initialization problems from the use of
+ICU.
+
+There are, however, some significant limitations for applications that make use
+of ICU at C++ static initialization time:
 
-(For some platforms, the configure option `--enable-auto-cleanup` (or
-defining the option `UCLN_NO_AUTO_CLEANUP` to 0) will add code which
-automatically cleans up ICU when its shared library is unloaded. See comments in
-`ucln_imp.h`)
+1.  `u_setMemoryFunctions()` and the data locating functions, if needed, must
+still be called before any other use of ICU. Which includes any use during the
+construction of static objects.
+
+2.  `u_cleanup()` can only be called after all ICU objects have been destructed
+(C++) or closed (plain C APIs). Meaning that `u_cleanup()` cannot be called
+directly just before the application's `main()` function returns. It must either
+be skipped, or called from some other static destructor that runs after all use
+of ICU is done. Skipping `u_cleanup()` is not desirable; doing so will cause leak
+detection tools to falsely report problems with the ICU library.
+
+3.  Destruction of static objects that are scoped to a code block. These, by the
+conventions of C++, are lazily initialized when the code block is first entered,
+so there are no issues during static initialization. But object destruction
+happens when the program terminates, leaving the problem of where to call
+`u_cleanup()`, as discussed above.
+
+#### Dynamically Loading and Unloading ICU
```

Comment:
I had a comment suggesting such a section, and here it is :-)
@jefgen do you want to add something here?

Reply:
Thanks! -- Though I'm not sure if I really have very much to add though. (cc'ing @daniel-ju as well, in case I'm missing something whilst OOF. :) )

Perhaps the following points maybe (?):
- The loading and unloading of ICU should *never* be done inside [DllMain](https://docs.microsoft.com/en-us/windows/win32/dlls/dllmain). (Loading one of the ICU libraries can cause other libraries or files to be loaded, leading to potential dead-lock.)
- When dynamically loading or unloading ICU, after closing and deleting all ICU objects and services, you should _always_ call `u_cleanup()` before unloading, otherwise you will leak OS Critical Sections. (Or perhaps maybe just say that you will leak things?).

Though the text already says to always call `u_cleanup()` -- so perhaps the second point is redundant.

- Sort of related to the comment here https://github.com/unicode-org/icu/pull/1513#discussion_r544952547 -- if you are using ICU with a data _file_ instead of a data _dll_ then you may want to call `ulocdata_getCLDRVersion()` after loading the ICU libraries before using other APIs to ensure the data file is actually loaded.

Reply:
> you should always call u_cleanup() before unloading, otherwise you will leak OS Critical Sections. (Or perhaps maybe just say that you will leak things?).

its worth adding that not just dynamic memory might be leaked.